adam/perception/perception_graph.py

```diff
@@ -313,111 +328,113 @@ def from_schema(object_schema: ObjectStructuralSchema) -> "PerceptionGraphPatter
+        islands = list(isolates(perception_graph.graph))
+        perception_graph.graph.remove_nodes_from(islands)
+
+        # Use from_graph code to translate Graph to Pattern
+        return PerceptionGraphPattern.from_graph(perception_graph=perception_graph)
 
     @staticmethod
-    def _translate_schema(
-        object_: Any,
-        object_schema: ObjectStructuralSchema,
-        graph: DiGraph,
-        schema_node_to_pattern_node: Dict[Any, "NodePredicate"],
-    ) -> None:
-        def map_node(node: Any) -> NodePredicate:
-            # we need to do this because multiple sub-objects with the same schemata
-            # (e.g. multiple tires of a truck) will share the same Geon, Region, and Axis objects.
-            # We therefore need to "scope" those objects to within e.g. a single tire.
-            # If any new types of predicates are added below,
-            # be sure to check the predicate implementation is aware of the tuple-wrapping.
-            key: Any
-            if isinstance(node, SubObject):
-                key = node
-            else:
-                key = (object_, node)
-
-            if key not in schema_node_to_pattern_node:
-                if isinstance(node, GeonAxis):
-                    schema_node_to_pattern_node[key] = AxisPredicate.from_axis(node)
-                elif isinstance(node, Geon):
-                    schema_node_to_pattern_node[key] = GeonPredicate.exactly_matching(
-                        node
-                    )
-                elif isinstance(node, Region):
-                    schema_node_to_pattern_node[key] = RegionPredicate.matching_distance(
-                        node
-                    )
-                elif isinstance(node, SubObject):
-                    schema_node_to_pattern_node[key] = AnyObjectPerception(
+    def from_graph(perception_graph: PerceptionGraph) -> "PerceptionGraphPattern":
+        """
+        Creates a pattern for recognizing an object based on its *perception_graph*.
+        """
+        pattern_graph = DiGraph()
+        PerceptionGraphPattern._translate_graph(
+            perception_graph=perception_graph.graph, pattern_graph=pattern_graph
+        )
+        return PerceptionGraphPattern(pattern_graph)
+
+    @staticmethod
+    def _translate_graph(perception_graph: DiGraph, pattern_graph: DiGraph) -> None:
+        perception_node_to_pattern_node: Dict[Any, "NodePredicate"] = {}
+        # perception graph -> perception graph pattern
+
+        def map_node(node: Any) -> "NodePredicate":
+
+            key = node
+            if key not in perception_node_to_pattern_node:
+                if isinstance(node, tuple):
+                    node = node[0]
+                    if isinstance(node, Geon):
+                        perception_node_to_pattern_node[
+                            key
+                        ] = GeonPredicate.exactly_matching(node)
+                    elif isinstance(node, Region):
```

Comment:
Add an `else` which throws an exception. That way, if someone adds something we don't expect, we will find out about it quickly.